src/pam_google_authenticator.c

```diff
@@ -1493,6 +1513,125 @@ static int check_timebased_code(pam_handle_t *pamh, const char*secret_filename,
   return 1;
 }
 
+/* 
+ * Add a 'config' variable that says we logged in from a particular place
+ * at a particular time.  Only remembers the last 10 logins, which 
+ * are replaced in LRU order.
+ *
+ * Returns 0 on success.
+ */
+int
+update_logindetails(pam_handle_t *pamh, const Params *params, char **buf) {
+  const char *rhost = get_rhost(pamh, params);
+  const time_t now = get_time();
+  time_t oldest = now;
+  int oldest_index = 0;
+  int i;
+  char *line;
+  unsigned long when;
+  char name[] = "LAST ";
+  char *host;
+
+  if (rhost == NULL) {
+    return -1;
+  }
+
+  for (i = 0; i < 10; i++) {
+    name[4] = i + '0';
+    line = get_cfg_value(pamh, name, *buf);
+    if (line == &oom) {
+      /* Fatal! */
+      return -1;
+    }
+
+    if (!line) {
+      /* Make first empty line the oldest */
+      if (oldest) {
+        oldest_index = i;
+        oldest = 0;
+      }
+      continue;
+    }
+
+    if (sscanf(line, " %m[0-9a-fA-F:.] %lu ", &host, &when) != 2) {
```

Comment:
%m doesn't seem to be super portable. It's not supported on latest OpenBSD (according to my test), for example, so I suspect problems with other BSDs.

Reply:
Fixed.  I wanted to have to avoid thinking about buffer overflow if someone's edited the file; but if it's not going to be portable, ...